workflow-core/src/commonMain/kotlin/com/squareup/workflow1/StatefulWorkflow.kt

```diff
@@ -73,7 +75,381 @@ public abstract class StatefulWorkflow<
+      E2,
+      E3,
+      E4,
+      E5,
+      E6,
+      E7,
+      E8,
+      E9
+      > safeEventHandler(
+      name: String = "safeEventHandler",
+      crossinline onFailedCast: (name: String, type: KClass<*>, state: StateT) -> Unit =
+        ::defaultOnFailedCast,
+      crossinline update:
+      WorkflowAction<@UnsafeVariance PropsT, StateT, @UnsafeVariance OutputT>.Updater.(
+        currentState: CurrentStateT,
+        e1: E1,
+        e2: E2,
+        e3: E3,
+        e4: E4,
+        e5: E5,
+        e6: E6,
+        e7: E7,
+        e8: E8,
+        e9: E9
+      ) -> Unit
+    ): (E1, E2, E3, E4, E5, E6, E7, E8, E9) -> Unit {
+      return eventHandler(
+        { name }
+      ) { e1: E1, e2: E2, e3: E3, e4: E4, e5: E5, e6: E6, e7: E7, e8: E8, e9: E9 ->
+        CurrentStateT::class.safeCast(state)
+          ?.let { currentState -> this.update(currentState, e1, e2, e3, e4, e5, e6, e7, e8, e9) }
+          ?: onFailedCast(name, CurrentStateT::class, state)
+      }
+    }
+
+    public inline fun <
+      reified CurrentStateT : StateT & Any,
+      E1,
+      E2,
+      E3,
+      E4,
+      E5,
+      E6,
+      E7,
+      E8,
+      E9,
+      E10
+      > safeEventHandler(
+      name: String = "safeEventHandler",
+      crossinline onFailedCast: (name: String, type: KClass<*>, state: StateT) -> Unit =
+        ::defaultOnFailedCast,
+      crossinline update:
+      WorkflowAction<@UnsafeVariance PropsT, StateT, @UnsafeVariance OutputT>.Updater.(
+        currentState: CurrentStateT,
+        e1: E1,
+        e2: E2,
+        e3: E3,
+        e4: E4,
+        e5: E5,
+        e6: E6,
+        e7: E7,
+        e8: E8,
+        e9: E9,
+        e10: E10
+      ) -> Unit
+    ): (E1, E2, E3, E4, E5, E6, E7, E8, E9, E10) -> Unit {
+      return eventHandler(
+        { name }
+      ) { e1: E1, e2: E2, e3: E3, e4: E4, e5: E5, e6: E6, e7: E7, e8: E8, e9: E9, e10: E10 ->
+        CurrentStateT::class.safeCast(state)
+          ?.let { currentState ->
+            this.update(currentState, e1, e2, e3, e4, e5, e6, e7, e8, e9, e10)
+          }
+          ?: onFailedCast(name, CurrentStateT::class, state)
+      }
+    }
+  }
+
+  /**
+   * Like [action], but no-ops if [state][WorkflowAction.Updater.state] has
+   * changed to a different type than [CurrentStateT] by the time [update] fires.
+   *
+   *    private fun stopPlaying(
+   *      game: CompletedGame
+   *    ) = safeAction<Playing>("stopPlaying") { currentState ->
+   *      state = when (game.ending) {
+   *        Quitting -> MaybeQuitting(currentState.playerInfo, game)
+   *        else -> GameOver(currentState.playerInfo, game)
+   *      }
+   *    }
+   *
+   * This is not an uncommon case. Consider accidental rapid taps on
+   * a button, where the first tap event moves the receiving [StatefulWorkflow]
+   * to a new state. There is no reason to expect that the later taps will not
+   * fire the (now stale) event handler a few more times. No promise can be
```

Comment:
> the (now stale) event handler

Is it confusing to use the term `event handler` here? The WorkflowAction created by `safeAction` is probably being used to handle an event, but it might be better to use a different term that isn't easily confused with the `eventHandler` method. How about `the (now stale) [update] lambda`? That change would also work well in the `safeEventHandler` documentation.